enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,427 @@
+
+As mentioned in the above section, degrading will only happen when the patching of the MachineSet fails. This is likely due to a temporary API server outage and will resolve itself without user intervention. The degrade condition is calculated at the end of a sync loop. In the case of multiple such failures within a single sync loop, the message for the degrade will be accumulated to include the `MachineSets` associated with all the failures. 
+
+#### Reverting to original bootimage
+
+The proposal will introduce a CR, `MachineSetBootImageHistory` to store the boot image history associated with a given machineset. By providing this CR and accompanying documentation, the user will be able to restore their machinesets to an earlier state if they wish to do so. 
+
+### Workflow Description
+
+It is important to note that there would be two "opt-in" knobs while this feature is under TechPreview. The user would first have to turn on the feature gate, and then the opt-in mechanism. The secondary knob is necessary as some customers may want to keep their boot images static when this feature leaves TechPreview.
+
+See the API extension section for examples of how this feature can be turned on and off. 
+
+#### Variation and form factor considerations [optional]
+
+Any form factor using the MCO and `MachineSets` will be impacted by this proposal. So case by case:
+- Standalone OpenShift: Yes, this is the main target form factor.
+- microshift: No, as it does [not](https://github.com/openshift/microshift/blob/main/docs/contributor/enabled_apis.md) use `MachineSets`.
+- Hypershift: No, Hypershift does not have this issue.
+
+##### Supported platforms
+
+The initial release(phase 0) will support GCP. In future releases, we will add in support for remaining platforms as we gain confidence in the functionality and understand the specific needs of those platforms. For platforms that cannot be supported, we aim to atleast provide documentation to perform the boot image updates manually. Here is an exhaustive list of all the platforms:
+
+- gcp
+- aws
+- azure
+- alibabacloud
+- nutanix
+- powervs
+- openstack
+- vsphere
+- baremetal
+- libvirt
+- ovirt
+- ibmcloud
+
+This work will be tracked in [MCO-793](https://issues.redhat.com/browse/MCO-793).
+
+##### Cluster API backed machinesets
+
+As the Cluster API move is impending(initial release in 4.16 and default-on release in 4.17), it is necessary that this enhancement plans for the changes required in an CAPI backed cluster. Here are a couple of sample YAMLs used in CAPI backed `Machinesets`, from the [official Openshift documentation](https://docs.openshift.com/container-platform/4.14/machine_management/capi-machine-management.html#capi-sample-yaml-files-gcp).
```

Comment:
Do we need to have a distinction in the API here that shows Machine API Machines and Cluster API Machines in a separate light? Or do we expect to select all MachineSets independent of the API group they are in? 🤔

Reply:
Just to clarify, when you say in the API - are you taking about the configuration API for machinesets(via label selector) we are currently working on? I figured that could work with both form factors since it is using label selectors as long as we wire up the MSBIC to look for both. If it's about the boot image history CRD, I agree, a distinction would be helpful

Reply:
My expectation here is that eventually Cluster API will take over, and all MachineSets will be managed via Cluster API. However, in the interim period, there will be both Machine API and Cluster API MachinesSets, and in some cases, those will represent the same groups of Machines.

The reason I'm thinking to have it so that it's explicitly called out in the `MachineConfiguration` API is so that, from a MCO perspective, it's obvious which path you need to follow to update the images. You'll likely have two separate and very different code paths for updating the images, which makes me think, if we leave this to the user to configure, they'll be able to make that distinction. I've moved this MachineSet to CAPI management now, so I'll update my config to point to Cluster API instead of Machine API.